### ISSUE TITLE
[Feature request] Cohere Transcribe engine for macOS (Arabic and 14 languages, ASR + forced alignment)

### ISSUE THREAD

**Issue by muaz978** [SubtitleEdit/subtitleedit#12290]
### Summary

I have built and benchmarked a new macOS speech to text engine based on Cohere Labs' open source ASR models (Apache 2.0), and I would like your confirmation before opening the integration pull request. Everything is already prepared and tested; this issue is to agree on whether Subtitle Edit wants an engine with these requirements, and how you would like it structured.

Cohere Transcribe is a 2B parameter dedicated ASR model. There are two variants, both Apache 2.0:

- `CohereLabs/cohere-transcribe-arabic-07-2026`: Arabic and English, tuned for Arabic dialects and Arabic/English code switching.
- `CohereLabs/cohere-transcribe-03-2026`: 14 languages (English, French, German, Italian, Spanish, Portuguese, Greek, Dutch, Polish, Mandarin, Japanese, Korean, Vietnamese, Arabic).

The model emits plain text with no timestamps, so the engine supplies timing itself: Silero VAD finds the speech regions, Cohere transcribes each region, and Meta's MMS forced aligner (CTC, supports 1100+ languages) aligns the text back onto the audio for word level timestamps, which then feed the same cue builder used by the MLX Whisper Mac engine.

### Why: benchmarks on Apple Silicon (M-series)

All runs on the same audio, MLX large-v3 with beam size 5 as the reference.

| Test | Cohere | MLX large-v3 + beam 5 | Result |
|---|---|---|---|
| Arabic, real TV episode, 3 min | 72 s, clearly more accurate | 115 s | Cohere wins on accuracy and speed |
| English, clean studio clip, 25 s | 15 s, flawless | 10 s, flawless | Parity, MLX marginally faster |
| English, hard 1949 radio drama, 3 min | 41 s, full content | 108 s, full content | Comparable text, Cohere much faster |

On the Arabic episode Cohere produced the correct `للمثول`, `سيدتي`, `زميلاتكن`, `سيقتفون`, `أسرار`, `سرقة`, `إيقاعك`, `نشاطه` where MLX produced `المجول`, `سيادتي`, garbled text, `سيقطفون`, `إسرار`, `صرقة`, and so on. On the hard English drama Cohere avoided a Whisper style hallucination loop that MLX fell into near the end, and read `prowled the streets` correctly where MLX had `follow the streets`. On easy English the two are indistinguishable.

Net: Cohere is a strong Arabic specialist that is competitive on English, with a speed advantage and less tendency to hallucinate on difficult audio. For the Arabic and MENA user base this is a meaningful upgrade over Whisper.

### Requirements and honest caveats

This engine is heavier than the existing ones, which is exactly why I am asking first:

- Python 3.10 or newer (Cohere requires transformers 5.4 or newer; the current Mac engines run on 3.9). Subtitle Edit's Python discovery would need to prefer a 3.10+ interpreter for this engine.
- PyTorch plus transformers, and about 5 GB of model downloads (Cohere 2B is roughly 4 GB, the MMS aligner ONNX is about 1.2 GB), cached after first use.
- The Cohere models are Apache 2.0 but gated on Hugging Face, so each user accepts the model terms once and logs in with their own free Hugging Face token. No key is bundled.
- macOS Apple Silicon (Metal via MPS). It also runs on CPU, slower.

### Install and run (will go in the PR and the README / setup notes)

```
# 1. A Python 3.10+ interpreter (Homebrew shown)
brew install python@3.12

# 2. Dependencies
python3.12 -m venv se-cohere && source se-cohere/bin/activate
pip install "transformers>=5.4.0" torch huggingface_hub soundfile librosa \
            sentencepiece protobuf accelerate ctc-forced-aligner unidecode faster-whisper

# 3. Accept the model terms once at
#    https://huggingface.co/CohereLabs/cohere-transcribe-arabic-07-2026
#    (or the 14 language cohere-transcribe-03-2026), then log in with a free read token
hf auth login

# 4. Subtitle Edit calls the helper like the other engines, for example
python3.12 cohere_transcribe.py --audio input.wav --language ar \
    --output-dir out --output-format srt
```

First run downloads the models; later runs are offline. The helper prints `[MM:SS --> MM:SS] text` per cue for the live console, exactly like the MLX helper.

### What is ready

The engine helper script (about 530 lines, reusing the MLX helper's cue builder), the VAD chunking, the MMS alignment, and the SRT output are all written and tested end to end on macOS. If you are open to an engine with the requirements above, I will open the integration pull request (engine class, wiring into the speech to text engine list, Python 3.10+ discovery, settings, and download hints), mirroring the MLX Whisper Mac engine, with the benchmark evidence and screenshots attached.

Please let me know if you want this, and whether you would prefer the Arabic specialist model, the 14 language general model, or both selectable.

**Comment by muaz978** [comment]
@niksedk for visibility, one clarification on scope:

To be explicit about scope: I intend the engine to support all 14 languages of the general model, not only Arabic. The plan is a model selector, the Arabic specialist for the best Arabic quality, and the 14 language general model (English, French, German, Italian, Spanish, Portuguese, Greek, Dutch, Polish, Mandarin, Japanese, Korean, Vietnamese, Arabic) for everything else, both selectable in the engine settings. The MMS forced aligner already covers 1100+ languages, so the word level timestamps and cue building work for every one of those languages without extra work. So this is not an Arabic only addition, it is a multilingual engine that happens to be strongest on Arabic.

**Comment by niksedk** [comment]
We already have Cohere via Crisp ASR - and calling python is something I would rather avoid as it can be very tricky.

**Comment by muaz978** [comment]
You are right, and thank you for pointing me to Crisp ASR. I checked it: Crisp ASR Cohere already runs Cohere locally through the native GGUF binary, with its own forced aligner and SRT output, and no Python. That is clearly the better path than the Python engine I proposed, so I will drop the Python idea.

There is one worthwhile gap though. Crisp ASR Cohere uses the general `cohere-transcribe-03-2026` model. For Arabic there is a dedicated model, `cohere-transcribe-arabic-07-2026`, tuned for Arabic dialects and Arabic/English code switching. On Cohere's own Open Universal Arabic ASR leaderboard it is a large step better on Arabic (about 25.9 WER versus about 30.7 for the general model), and in my own side by side tests on real Arabic TV material it was clearly more accurate than the general model.

The good part: that Arabic model is already published as GGUF by the same author whose general GGUFs Crisp ASR uses (`cstr/cohere-transcribe-arabic-07-2026-GGUF`, same quant set: f16, q4_k, q8_0). So it can be added as an extra model option in the existing Crisp ASR Cohere engine with no new dependency, no Python, and no new infrastructure, just a few model entries next to the existing q4_k to q8_0 ones. Users who work in Arabic pick the Arabic model; everyone else keeps the general one.

If you are open to it I will prepare that small addition to Crisp ASR Cohere. It gives Arabic and MENA users a real accuracy gain while staying entirely on the native path you prefer.

**Comment by muaz978** [comment]
Before we close this off, one thing I wanted to make sure is on your radar: the engine we prepared is more capable than transcription alone. It pairs Cohere with an MMS forced aligner that supports over 1100 languages for alignment, so it can produce accurate word and segment timing well beyond the languages we transcribe today, and it handles long files with its own VAD based windowing. In my tests that combination gave noticeably tighter time codes, which is the part that is usually hardest to get right.

Two questions, so I do not push anything you do not want:

1. Are you sure you would not want it added at all, even as an optional engine that only runs when the user explicitly picks it? If the concern is the Python dependency, I can keep it fully self contained and opt in, with no effect on anyone who does not select it.

2. Independently of the above, would you be open to adding the Arabic specialist model to the existing native Crisp ASR Cohere path? It is a dedicated Arabic and English model and is clearly more accurate for Arabic than the general one, with no Python involved at all. I have it staged and can open a small PR whenever you give the word.

Either way, thanks for taking the time to look at this.
@niksedk

**Comment by niksedk** [comment]
Thanks - I've gone ahead and added both to the existing native Crisp ASR Cohere engine: #12383

* **Arabic**: `cstr/cohere-transcribe-arabic-07-2026-GGUF` (q4_k / q8_0 / f16), exactly as you suggested - same uploader as the general Cohere GGUFs we already ship.
* **Japanese**: while looking I also found `efwkjn/cohere-asr-ja`, a Japanese fine-tune of the same `cohere-transcribe-03-2026` base, which is well ahead of the stock model on Japanese CER. Added the GGUF quants of that as well (q4_k / q6_k / q8_0 / f16).

So no Python, no new engine, just more model entries - which is the path I prefer.

On the forced aligner point: that part of your proposal is the genuinely interesting one.
Might be worth opening an issue over at https://github.com/CrispStrobe/CrispASR

**Comment by muaz978** [comment]
Thank you, this is a great outcome. The Arabic model is exactly what I hoped for, and adding the Japanese fine tune on top is a nice bonus for that audience. Keeping it all on the native GGUF path with no Python is clearly the right call.

I will take the forced aligner idea to the CrispASR repo as you suggested, since that is where it belongs. Appreciate you looking at all of this and folding it in so quickly.

**Comment by muaz978** [comment]
Following up with measured numbers on the Arabic model you added in #12383.

I benchmarked it against the general Cohere model on real Arabic material (Sayf Ibn Dhi Yazan E16, classical MSA), using Crisp ASR with `--vad` and scoring WER/CER against a human-made reference SRT.

![WER chart](https://raw.githubusercontent.com/muaz978/cohere-arabic-asr-benchmark/main/wer_chart.png)

Aggregate over 620 reference words, both with VAD:

| Model | WER | CER |
| --- | --- | --- |
| general `cohere-transcribe` | 20.6% | 21.1% |
| arabic `cohere-transcribe` (#12383) | 10.3% | 5.9% |

The Arabic model roughly halves WER (about 50% relative) and is much more robust: on one span the general model collapsed to 73% WER (a hallucination cascade) while the Arabic model held at 3.3%. So the addition is a clear win for Arabic and MENA users, thanks for merging it.

One engine note that came out of this: VAD is essential here. Without `--vad` the general model drifted to 52% WER on a 100 second span; with it, 20%. Might be worth defaulting or documenting for the Cohere backend.

Full per-span tables, the chart, and reproducible scripts: https://github.com/muaz978/cohere-arabic-asr-benchmark

**Comment by niksedk** [comment]
Both models were added to the existing native Crisp ASR Cohere engine in #12383 (9cce3ec94) — Arabic (`cstr/cohere-transcribe-arabic-07-2026-GGUF`) plus the Japanese fine-tune, on the native GGUF path with no Python dependency. @muaz978 benchmarked the Arabic model afterwards and the WER/CER numbers look good.

Closing this one — the forced-aligner part is being taken up in the CrispASR repo, and general alignment interest is tracked in #9256.
